### ISSUE TITLE
Unable to download arbitrary data entities

### ISSUE THREAD

**Issue by melissachang** [DataBiosphere/saturn-ui#579]
I think this belongs in https://github.com/broadinstitute/firecloud-orchestration, but that appears to have Issues disabled.

FireCloud only had sample and participant entities. With the new Data Explorer -> Saturn flow, we are creating entities of different types (BigQuery_table, cohort).

Repro:
- https://test-data-explorer.appspot.com
- Click on bottom right button
- Select workspace, click Import
- Click on BigQuery_table under Data Model
- Click Download
- Get error "FireCloudException: Error generating entity download: Unknown entity type: BigQuery_table"

cc @malathi13

**Comment by bradtaylor** [comment]
Making this work entails a change to either Rawls or FC-orch, and/or waiting for entity service. Forward to Content&Discovery team

**Comment by ahaessly** [comment]
@bradtaylor can you give me access to the test workspace?

**Comment by bradtaylor** [comment]
@ahaessly shared this with you (table referencing a public bigquery. Appears to happen for any non-participant/sample/pair/set data table)
https://bvdp-saturn-prod.appspot.com/#workspaces/broad-firecloud-dsde/bigquery/data

**Comment by ahaessly** [comment]
In the code for export, it treats collection types differently than non-collection types. it tries to determine if the entity type passed in is a collection type or not and that is where the "unknown entity type" occurs.

we can probably do some minor refactoring in orchestration if the data that is being exported is not a collection type. when creating the tsv, the code also creates the name for the plural even if it is not used - that would also fail.

bigger question. will we need to also support collection types that are not part of our model? if so, then the solution is not so straightforward. we would need some way to determine for the entity type and field if it represents a collection.

**Comment by melissachang** [comment]
It's complicated. The cohort entity is conceptually collection of participants, but it's not implemented as a list of participant ids. (It's a SQL query.)

Andrea, I'll schedule a meeting with you to discuss. Let me know if anyone else wants to attend.

We can also discuss how in Q4, we would like Saturn and/or EntityService to materialize cohort into a table.

**Comment by ahaessly** [comment]
@bradtaylor sorry, that comment was cryptic - too implementation specific. basically, with some very minor refactoring we can export non-collection types that are not part of the firecloud model. using a convention like _set, with some additional minor refactoring, i believe we could support export of a set of types also not supported by the firecloud model.

**Comment by kleighwe** [comment]
@kleighwe will schedule a meeting to discuss with @bradtaylor @panentheos @ahaessly and Moran.

**Comment by melissachang** [comment]
I chatted with Andrea. For this particular bug (which is not urgent), the fix is a small code change (I think in EntityService).

If the meeting is to discuss materialization, please invite me. :)

**Comment by kleighwe** [comment]
Will do, @melissachang! @bradtaylor- does the above comment answer our questions? Still want me to find a good time for us all to check in?

**Comment by kleighwe** [comment]
Andrea is working on this. Jira Ticket: https://broadinstitute.atlassian.net/secure/RapidBoard.jspa?rapidView=476&view=detail&selectedIssue=GAWB-3847

**Comment by ahaessly** [comment]
i've got some code ready for review if anyone is interested

**Comment by kleighwe** [comment]
@bradtaylor this should be unblocked! The work Andrea did is on Alpha.

**Comment by bradtaylor** [comment]
See email for endpoint. It may point to orchestration. It may be worth asking if there's a deeper endpoint, but the team concurs orchestration is acceptable for now
